src/pip/_internal/network/download.py

```diff
@@ -163,20 +173,40 @@ def __init__(
         # type: (...) -> None
         self._session = session
         self._progress_bar = progress_bar
+        self._tmpdir = TempDirectory(kind='unpack', globally_managed=True)
 
-    def __call__(self, link):
-        # type: (Link) -> Download
+    def _download(self, link, location):
+        # type: (Link, str) -> File
+        url, sep, checksum = link.url.partition('#')
+        response = self._session.get(url, headers=HEADERS, stream=True)
         try:
-            resp = _http_get_download(self._session, link)
+            raise_for_status(response)
         except NetworkConnectionError as e:
             assert e.response is not None
             logger.critical(
                 "HTTP error %s while getting %s", e.response.status_code, link
             )
             raise
 
-        return Download(
-            resp,
-            _get_http_response_filename(resp, link),
-            _prepare_download(resp, link, self._progress_bar),
-        )
+        chunks = _prepare_download(response, link, self._progress_bar)
+        filename = _get_http_response_filename(response, link)
+        file_path = os.path.join(location, filename)
+
+        with open(file_path, 'wb') as content_file:
+            for chunk in chunks:
+                content_file.write(chunk)
+        return file_path, response.headers.get('content-type', '')
+
+    def __call__(self, link, location=None, hashes=None):
```

Comment:
Previously this class and function call did essentially 1 thing: download a file with progress. That was a good thing because it could be understood as a single small chunk, passed around, and used without much effort. When we add more responsibilities it gets harder to understand and maintain.

Instead of pulling functionality from RequirementPreparer into this, and then invoking this from the new resolver, what if we:

1. move the functionality from RequirementPreparer somewhere else and invoke it from RequirementPreparer and the new resolver, or
2. (preferred) move downloading out of RequirementPreparer all together, and do it either:
    1. internally in AbstractDistribution/Candidate when required to satisfy some part of processing (transparently to the caller), or
    2. in a common place after resolving has happened, if still needed

Reply:
> (preferred) move downloading out of RequirementPreparer all together, and do it either:

I haven't looked at this PR yet, but IIUC, this was the conclusion from what @McSinyx and I discussed. I do think I agree that `Downloader` might not be the best place to put this though. :)

Reply:
On current master, network.download.Downloader only initiates the download and passes the lazy response to the preparation code which then performs the actual download (to a file).  What 744896f3508174c74120172a5b529107674d6175 aims to do is making sure that `Downloader.__call__` will perform the download, with addition to hash checking to make the interface consistent when caching is involved.  In the way I see it, Downloader should be able to act analogous to wget, which is given an URL/Link, optional output target (`-O`).  I assume *more responsibilities* refers to the hash checking.

The reason I placed the call to Downloader directly in `resolve` is that it is easier to ensure that such call is thread-safe than a higher-level wrapper like a method in `RequirementPreparer`.   It would also be more trivial to handle the progress bar for a batch of downloads directly within the Downloader than extracting it from the preparer.

If I understand correctly, then RequirementPreparer seems to be the preparer for candidates of the new resolver, and downloading something happens to be part of the work for some candidates, i.e. RequirementPreparer sometimes need to download to be able to construct an (Abstract)Distribution to give to the candidate, so I'm not sure if I understand the model (2.i) is referring to.

Regarding (2.ii), I completely agree that it would be really nice if we can have this call outside of the resolver code.  However, in a discuss with @pradyunsg, we came to a conclusion that InstallRequirement is currently to tangled with the preparation code (or that it logically requires the underlying file to exists), so that we will move the call out once we reworked RequirementSet.  @chrahunt, I'm not sure if I've correctly understood and addressed your points since my head is buzzing  as I try to visualize this but I am able to deliver some reasonings.  Do they change your mind about 744896f3508174c74120172a5b529107674d6175 or do you have any further suggestions?

Reply:
> @pradyunsg, we came to a conclusion that InstallRequirement is currently to tangled with the preparation code (or that it logically requires the underlying file to exist), so that we will move the call out once we reworked RequirementSet.

The issue is that the code that gets run *after* `Resolver.resolve` returns, is closely tied to the assumption that the underlying file exists. We don't need for reworking of RequirementSet, but rather, we need to cleanly handle the "I need to be fully downloaded" state without needing to touch the old resolver code.

Reply:
@McSinyx I think some of my concern was an extension of my original comments in #8532, which I did not follow up on in #8588 (sorry). I played with it a bit and came up with #8685, which I think we can use as a pretty clean base for your download work. Please take a look!

Regarding the downloader, I think we can probably achieve the desired effect by creating a new function or class that uses it as-is instead of putting more logic into it. Since our fast-deps feature isn't enabled when doing hash checking, the wrapper won't need any of that stuff.

Reply:
> We don't need for reworking of RequirementSet, but rather, we need to cleanly handle the "I need to be fully downloaded" state without needing to touch the old resolver code.

Thank you for the correction.

> I played with it a bit and came up with 8685, which I think we can use as a pretty clean base for your download work. Please take a look!

Thanks, I haven't read it carefully but the resulting hook seems really nice!

> Regarding the downloader, I think we can probably achieve the desired effect by creating a new function or class that uses it as-is instead of putting more logic into it.  Since our fast-deps feature isn't enabled when doing hash checking, the wrapper won't need any of that stuff.

This sounds good to me too.  @pradyunsg said that you two had a discussion on this and I assume this is the consensus that came out of it.

Reply:
I was thinking about the hashing part, and IMHO apparently we should check them, just to make sure that a wheel is not corrupted during the download process.

Reply:
> I was thinking about the hashing part, and IMHO apparently we should check them, just to make sure that a wheel is not corrupted during the download process.

If we're in "require-hashes" mode, doing partial downloads does not make sense, since we're only permitted to trust artifacts/files with that specific hash. And the only way to compute that hash is to download the entire file. Since we're going to download the entire file (and we know that if you have hashes, you also have performed dependency resolution already), we should not use partial downloads/fast-deps when hash-checking mode is enabled.

Reply:
I think I didn't communicate this right :stuck_out_tongue:  Let me provide more context (or at least the context I'm assuming):
* We want a routine taking multiple URLs and download them in parallel, with UI taken into account
* Before that's done, let's have a `map`/`for` that call something download of each of them, and that something should be thread-safe
* It'll be difficult to keep all preparations thread-safe, even it is at the moment (I think) https://github.com/pypa/pip/blob/8b838ebb89c521cd92dd7a1c7cec075c3b30d72f/src/pip/_internal/operations/prepare.py#L493-L498
* Rather, let what to be parallelized to be solely downloading the file to (usually) a temp dir and let the preparation code use it as cache latter.  Both (after download and when verifying the cache) needs hash checking because the file might corrupt for some reason.
* I'm not sure if it's the effect of `--no-cache-dir`, but some wheels just don't get cached within a session.  This doesn't seem to sufficient: https://github.com/pypa/pip/blob/370322eacfb698391d1f02c32e644cfb4de85eb0/src/pip/_internal/network/download.py#L63-L66 but this is: https://github.com/pypa/pip/blob/370322eacfb698391d1f02c32e644cfb4de85eb0/src/pip/_internal/operations/prepare.py#L290-L300
* If we use a common temp dir for all downloads, this can be down automatically and this common state really well suit `Downloader`.  Furthermore the similarity between Downloader.download_one (or something to be renamed from current `.__call__`) and Downloader.download_batch would be really nice.

I am truly sorry for not clarifying this earlier in a discussion point and ended up causing a decent amount of confusion :sweat_smile: